src/main/java/com/tester/processor/BaseProcessor.java

```diff
@@ -0,0 +1,296 @@
+      sb.append(method.getName());
+      sb.append(method.getDesc());
+      sb.append("\n");
+      sb.append("Line coverage: ");
+      sb.append(method.getLineCounter().getCoveredRatio() * 100);
+      sb.append("%");
+      sb.append("\n");
+      sb.append("Branch coverage: ");
+      sb.append(method.getBranchCounter().getCoveredRatio() * 100);
+      sb.append("%");
+      sb.append("\n\n");
+    }
+    sb.append("Here are the suspicious lines in the program not covered by any test in the " +
+            "test suite:\n");
+    Map<Integer, Boolean> lineCoverageStatus = summary.getLineCoverageStatus();
+    StringBuilder susLinesSb = new StringBuilder();
+    String filePath = getFilePath(MAIN, type, programFileName);
+    for (int l : suspiciousLines.get(type)) {
+      if (lineCoverageStatus.getOrDefault(l, false)) {
+        continue;
+      }
+      String programLine = getProgramLine(l, filePath);
+      susLinesSb.append("Line ");
+      susLinesSb.append(l);
+      susLinesSb.append(": ");
+      susLinesSb.append(programLine);
+      susLinesSb.append("\n");
+    }
+    if (susLinesSb.length() == 0) {
+      sb.append("None");
+    } else {
+      sb.append(susLinesSb);
+    }
+    return sb.toString();
+  }
+
+  void writeToFile(String type, String code) {
+    byte[] codeBytes = code.getBytes();
+    String packageName = getPackageName(type);
+    byte[] packageNameBytes = packageName.getBytes();
+    String filePath = getFilePath(TEST, type, testFileName);
+    try {
+      if (Files.exists(Paths.get(filePath))) {
+        Files.delete(Paths.get(filePath));
+      }
+      FileChannel channel = FileChannel.open(Paths.get(filePath), StandardOpenOption.CREATE,
+              StandardOpenOption.APPEND);
+      channel.write(ByteBuffer.wrap(packageNameBytes));
+      channel.write(ByteBuffer.wrap(codeBytes));
+      channel.force(true);
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+  }
+
+  private String getProgramLine(int lineNum, String filePath) {
+    int currLineNum = 1;
+    try {
+      BufferedReader reader = new BufferedReader(new FileReader(filePath));
+      String line = reader.readLine();
+      while (line != null && currLineNum < lineNum) {
+        line = reader.readLine();
+        currLineNum++;
+      }
+      return line;
+    } catch (IOException e) {
+      e.printStackTrace();
+      return "";
+    }
+  }
+
+  private String readProgram(String filePath) {
+    try {
+      List<String> lines = Files.readAllLines(Paths.get(filePath));
+      return String.join(System.lineSeparator(), lines);
+    } catch (IOException e) {
+      e.printStackTrace();
+      return "";
+    }
+  }
+
+   String getBaseName(String fileName) {
+    int dotIndex = fileName.lastIndexOf('.');
+    return (dotIndex == -1) ? fileName : fileName.substring(0, dotIndex);
+  }
+
+  protected abstract void writeTest2File(String code);
+
+  abstract String getFilePath(String dir, String type, String fileName);
```

Comment:
Question: What is the reasoning behind making these abstract? Cant this be a protected method that all subclasses use? Since in `MinerProcessor.java` you return null

Reply:
i implemented it in MinerProcessor on 1/16 afterwards